### ISSUE TITLE
Provide a singleton root logger

### ISSUE THREAD

**Issue by cafarm** [uber-go/zap#127]
Loggers typically provide a singleton root logger for convenience. I'm not seeing that in zap. 

I realize it's trivial to implement in application code, but I'm curious about the rationale behind the omission from zap. Is it planned and just not implemented yet?

**Comment by akshayjshah** [comment]
Once the base APIs are complete, we'll figure out how best to support this - I generally prefer injecting loggers over using a global, but I realize that this is a common pattern.

**Comment by akshayjshah** [comment]
As an implementation note, it'd be ideal if the singleton didn't require users to take a lock before logging.

**Comment by sectioneight** [comment]
How do we imagine configuring the singleton logger? Just require that `zap.Configure(opts...)` is called before it's available?

**Comment by sectioneight** [comment]
My two cents:

I think there's a lot of value to being able to call `<package>.Debug()` from code without necessarily having to thread through context. I'm not sure it should necessarily be `zap.Debug()` (like logrus provides) but along the theme of "make it easy to do the right thing" there are plenty of cases where context isn't really necessary but a properly configured (e.g. JSON-formatted, to-disk or to-stdout) logger **is** helpful.

If we don't implement this in `zap`, which I can understand the reasons not to, we're going to implement it in the service framework, meaning this library won't have control over it.

**Comment by akshayjshah** [comment]
@sectioneight Yeah, I'd imagined something like `zap.Configure(opts...)`. It'd be nice if the package `init` method configured a logger with some sane defaults though - perhaps a JSON encoder, logging info-level and above to stdout and stderr.

I'd definitely like this to be part of zap, rather than a separate package - it's a staple of  every other logging library, and users (quite reasonably) expect it.

**Comment by sectioneight** [comment]
Ok, take a look at the attached PR (#150) and see what you think

**Comment by rossnanop** [comment]
Hi Everyone,

New to zap, loving it so far.  I recently implemented this pattern in my own code and had an idea that has been working for me.  I don't have any opinions on where the singleton lives or what it is called, because ultimately it is not exposed via this API.  Let me know what you think!

```go
package logger

import (
	"sync"

	"github.com/uber-go/zap"
)

var singleton zap.Logger
var once sync.Once

// Init initializes a thread-safe singleton logger
// This would be called from a main method when the application starts up
// This function would ideally, take zap configuration, but is left out 
// in favor of simplicity using the example logger.
func Init() {
    // once ensures the singleton is initialized only once
	once.Do(func() {
		singleton = zap.New(
			zap.NewJSONEncoder(),
		)
	})
}

// Log a message at the given level with given fields
func Log(level zap.Level, message string, fields ...zap.Field) {
	singleton.Log(level, message, fields...)
}

// Debug logs a debug message with the given fields
func Debug(message string, fields ...zap.Field) {
	singleton.Debug(message, fields...)
}

// Info logs a debug message with the given fields
func Info(message string, fields ...zap.Field) {
	singleton.Info(message, fields...)
}

// Warn logs a debug message with the given fields
func Warn(message string, fields ...zap.Field) {
	singleton.Warn(message, fields...)
}

// Error logs a debug message with the given fields
func Error(message string, fields ...zap.Field) {
	singleton.Error(message, fields...)
}

// Fatal logs a message than calls os.Exit(1)
func Fatal(message string, fields ...zap.Field) {
	singleton.Fatal(message, fields...)
}

// ...
```

**Comment by ansel1** [comment]
Re: configuring sane defaults in init()

*Highly* recommend the singleton logger is configured as a no-op by default.  Let your main() or your test reconfigure it.  We've been using logxi for a year or so now, and the fact that it configures itself in its own init() function before we've had a chance to wrangle it to our own needs has been a constant source of issues.  

The main issue is that any package you bring in might start writing log messages to your singleton package logger in their own init() functions.  If your default package logger spits that out to stdout, there is *nothing* your code can do to stop it.  Yikes.

Personally, I like the pattern of creating package-level loggers for debug logging, either in static initializers or init() functions, rather than injecting loggers (it's hard to ensure you have a ref to an injected logger in some little standalone utility function where you want to drop in a trace log message).  But that only works if those pre-main() loggers are inert until main() tells them what to do and how to do it.

**Comment by akshayjshah** [comment]
Landed on dev in #271. @ansel1, made the defaults no-op loggers as you suggested.

**Comment by dfang** [comment]
@zythosec 

i copy your code to config/logger.go, and import logger to where i need to log

but i got this error 

```
../../go/pkg/mod/github.com/uber-go/zap@v1.10.0/stacktrace.go:28:2: use of internal package go.uber.org/zap/internal/bufferpool not allowed
```

**Comment by abhinav** [comment]
> @zythosec
> 
> i copy your code to config/logger.go, and import logger to where i need to log
> 
> but i got this error
> 
> ```
> ../../go/pkg/mod/github.com/uber-go/zap@v1.10.0/stacktrace.go:28:2: use of internal package go.uber.org/zap/internal/bufferpool not allowed
> ```

Hi, this issue is closed. If you have questions, please open a new issue.

That said, the problem in your case is that you're using the import path `github.com/uber-go/zap`, not `go.uber.org/zap`. The import path was changed before the 1.0 release.
